Add a component for Sisyphus Kinetic Art Tables

jkeljo opened the pull request:
## Description:

The [Sisyphus Kinetic Art Table](https://sisyphus-industries.com/) uses a steel ball to draw intricate patterns in sand, thrown into sharp relief by a ring of LED lights around the outside.

This component enables basic control of these tables through Home Assistant.

**Related issue (if applicable):** fixes #<home-assistant issue number goes here>

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation (if applicable):** home-assistant/home-assistant.github.io#5369

## Example entry for `configuration.yaml` (if applicable):
```yaml
sisyphus:

```

## Checklist:
  - [X] The code change is tested and works locally.
  - [X] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**

If user exposed functionality or configuration variables are added/changed:
  - [X] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
  - [X] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [X] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [X] New dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [X] New files were added to `.coveragerc`.

If the code does not interact with devices:
  - [ ] Tests have been added to verify that the new code works.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

jkeljo committed:
Add a component for Sisyphus Kinetic Art Tables

The [Sisyphus Kinetic Art Table](https://sisyphus-industries.com/) uses a
steel ball to draw intricate patterns in sand, thrown into sharp relief by a
ring of LED lights around the outside.

This component enables basic control of these tables through Home Assistant.

homeassistant commented:
Hi @jkeljo, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#14472).

Once you do that we will be able to review and accept this pull request.

Thanks!

jkeljo committed:
Fix lints

fabaff commented on homeassistant/components/light/sisyphus.py:
```diff
@@ -0,0 +1,66 @@
+        update_before_add=True)
+
+
+class SisyphusSwitch(Light):
```
Please add a docstring. Is required everywhere else too.

jkeljo committed:
Docstrings, other lints

syssi commented:
Some more lint issues copied from the logs of the travis builds above:

```
************* Module homeassistant.components.sisyphus
E: 42, 4: No name 'control' in module 'sisyphus' (no-name-in-module)
C: 46,10: Missing function docstring (missing-docstring)
E: 47,22: Module 'sisyphus' has no 'control' member (no-member)
C: 66,12: Variable name "ip" doesn't conform to snake_case naming style (invalid-name)
E: 66,24: Module 'sisyphus' has no 'control' member (no-member)
C: 73,10: Missing function docstring (missing-docstring)
************* Module homeassistant.components.media_player.sisyphus
R: 85,12: Unnecessary "else" after "return" (no-else-return)
E:158, 8: No name 'control' in module 'sisyphus' (no-name-in-module)
R:159, 8: Unnecessary "else" after "return" (no-else-return)
```

jkeljo committed:
More lints

jkeljo committed:
Yet more.

fabaff commented on homeassistant/components/sisyphus.py:
```diff
@@ -0,0 +1,81 @@
+        _LOGGER.debug("Connected to %s at %s", name, host)
+
+        hass.async_add_job(async_load_platform(
+            hass, 'light', 'sisyphus', {
```
Replace `'sisyphus'` with `DOMAIN`.

fabaff commented on homeassistant/components/sisyphus.py:
```diff
@@ -0,0 +1,81 @@
+            }, config
+        ))
+        hass.async_add_job(async_load_platform(
+            hass, 'media_player', 'sisyphus', {
```
Dito

fabaff commented on homeassistant/components/sisyphus.py:
```diff
@@ -0,0 +1,81 @@
+}, extra=vol.ALLOW_EXTRA)
+
+TABLES_SCHEMA = vol.Schema({
+    DOMAIN: [
```
Use `vol.Schema({..})` instead of a list.

fabaff commented on homeassistant/components/sisyphus.py:
```diff
@@ -0,0 +1,81 @@
+    ],
+}, extra=vol.ALLOW_EXTRA)
+
+CONFIG_SCHEMA = vol.Any(AUTODETECT_SCHEMA, TABLES_SCHEMA)
```
Use something like

```python
CONFIG_SCHEMA = vol.Schema({
    DOMAIN: vol.Any(AUTODETECT_SCHEMA, TABLES_SCHEMA)
}, extra=vol.ALLOW_EXTRA)
```

This would make the two different parts shorter and easier to read.

fabaff commented on homeassistant/components/sisyphus.py:
```diff
@@ -0,0 +1,81 @@
+            await add_table(ip_address)
+    else:  # TABLES_SCHEMA
+        for conf in table_configs:
+            if conf.get(CONF_HOST) is not None:
```
Can't be `None` it's required by the configuration validation.

jkeljo committed:
Feedback

jkeljo commented on homeassistant/components/media_player/sisyphus.py:
```diff
@@ -0,0 +1,212 @@
+    @property
+    def supported_features(self):
+        """Return the features supported by this table in its current state."""
+        if self.state == STATE_PLAYING:
```
I've changed it, but I have a question: how does one get the previous track button to be disabled when on the first track of the playlist, or the next track button to be disabled when on the last track?

jkeljo committed:
Lint

jkeljo committed:
Missed one piece of feedback

MartinHjelmare commented on homeassistant/components/media_player/sisyphus.py:
```diff
@@ -0,0 +1,202 @@
+    @property
+    def supported_features(self):
+        """Return the features supported by this table."""
+        return SUPPORT_VOLUME_MUTE \
```
Break out the features into a constant put at the module level and return that here.

MartinHjelmare commented on homeassistant/components/sisyphus.py:
```diff
@@ -0,0 +1,80 @@
+            await add_table(ip_address)
+    else:  # TABLES_SCHEMA
+        for conf in table_configs:
+            await add_table(conf.get(CONF_HOST), conf.get(CONF_NAME))
```
Don't use `dict.get` for required keys in the config dict.

MartinHjelmare commented on homeassistant/components/sisyphus.py:
```diff
@@ -0,0 +1,80 @@
+    async def close_tables(*args):
+        """Close all table objects."""
+        for table in tables.values():
+            await table.close()
```
We could use `asyncio.wait` here on a list of all tasks since the order to close the tables doesn't matter.

MartinHjelmare reviewed (changes requested):
Since the sisyphus-control library is using aiohttp for communication, it should accept a session from home assistant, so that a single session is used for the whole app.

MartinHjelmare commented on homeassistant/components/light/sisyphus.py:
```diff
@@ -0,0 +1,76 @@
+    async def async_added_to_hass(self):
+        """Add listeners after this object has been initialized."""
+        self._table.add_listener(
+            lambda: self.schedule_update_ha_state(False))
```
I looked into the library and the callback is called from the event loop. So we should use `async_schedule_update_ha_state` here instead.

jkeljo commented on homeassistant/components/media_player/sisyphus.py:
```diff
@@ -0,0 +1,202 @@
+            # otherwise there's a race condition by which this entity
+            # might not have had its hass field set, and thus
+            # the schedule_update_ha_state call will fail
+            self._table.add_listener(
```
Sorry about that...

jkeljo committed:
- Use async_added_to_hass in media player
- async_schedule_update_ha_state in listeners
- constants for supported features
- subscripting for required keys
- asyncio.wait
- update to sisyphus-control with passed-in session

jkeljo commented:
Had to rebase in order to bring in the new `sisyphus-control` with support for a passed-in session. Only the last commit is actually new.

MartinHjelmare commented on requirements_all.txt:
```diff
@@ -1243,6 +1243,9 @@ simplepush==1.1.4
 simplisafe-python==1.0.5
 
+# homeassistant.components.sisyphus
+sisyphus-control==2.0.0
```
Run `gen_requirements.py` again to update the requirements file. Then the build should pass.

jkeljo committed:
Update requirements

MartinHjelmare commented on homeassistant/components/media_player/sisyphus.py:
```diff
@@ -0,0 +1,197 @@
+        """Return the current state of the table; sleeping maps to off."""
+        if self._table.state in ["homing", "playing"]:
+            return STATE_PLAYING
+        elif self._table.state == "paused":
```
Replace `elif` with `if`.

MartinHjelmare commented on homeassistant/components/media_player/sisyphus.py:
```diff
@@ -0,0 +1,197 @@
+                return STATE_OFF
+
+            return STATE_PAUSED
+        elif self._table.state == "waiting":
```
See above.

MartinHjelmare reviewed (commented):
Small linting fixes needed, see Travis and below.

jkeljo committed:
lint

MartinHjelmare reviewed (approved):
Great!
